[JIT] Add Type support to the fuser, fuse more

t-vi opened the pull request:
This adds scalar type support to the fuser, both internally (instead of auto / assuming float) and for the inputs/outputs.
We can now fuse things with input / output of arbitrary scalar type, in particular comparisons and where work well. So it fixes #13384 by returning the right type tensor (and adds a test where byte and double tensors are returned).
The type inference is done by re-calling PropagateTensorShapeOnNode in the compilation, I would venture that it isn't prohibitively expensive compared to the actual compilation. (Propagation was fixed for where to return the second argument's type and amended to handle FusedConcat.)
I'm not sure how to add a check for the code generated by the fuser, but I am not sure we absolutely need to (we'd see if it is invalid / produces wrong results).

Thanks in particular to @apaszke, @fmassa, @mruberry for advice and encouragement! All the errors are my own.

I have discussed order of PRs briefly with @mruberry, if this goes in before he submits the PR, he graciously agreed to rebasing his, but I'd happily rebase, too.

t-vi committed:
Have PYTORCH_FUSION_DEBUG print C kernel source

- Move up handling the environment variable from CPU only to all
- Introduce two levels to be enabled with PYTORCH_FUSION_DEBUG=n:
  1: print C source
  2: print CPU assembly, too (previous effect of PYTORCH_FUSION_DEBUG)

t-vi committed:
Merge branch 'master' into pytorch_fusion_debug

t-vi committed:
fusiomania

t-vi committed:
Merge branch 'master' into fusion

t-vi committed:
shape/type analysis for fused concat (for fuser type analysis benefit)

t-vi committed:
Use output type of the right output instead of just the first

t-vi committed:
Cast operands of arithmetic operations to result type to better match PyTorch scalar casting

t-vi committed:
fuse aten::type_as

t-vi committed:
reenable a few fusion tests

t-vi committed:
add test for typing support and where

eplato commented:
Thanks!

t-vi committed:
reflect improved clamp backward fusion

t-vi commented:
Given that the asan (undefined behaviour through division by zero) failure isn't unique to this PR, I'd say the CI looks OK.

ngimel commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -257,8 +257,14 @@ void PropagateCatShape(Node * cat_node) {
   };
-  auto list_node = cat_node->namedInput(attr::tensors)->node();
-  if (list_node->kind() == prim::ListConstruct) {
+  if (cat_node->kind() == prim::FusedConcat) {
```
Is this necessary? `tensors` go out of scope right away, does `cat_node->inputs()` have some side effects? I would hope not.

ngimel commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -257,8 +257,14 @@ void PropagateCatShape(Node * cat_node) {
+		    ? cat_node
+                    : cat_node->namedInput(attr::tensors)->node());
+  if (list_node->kind() == prim::ListConstruct
+      || cat_node->kind() == prim::FusedConcat) {
```
IMO, would be slightly more readable if it's `list_node` here.

t-vi replied:
While I agree that `list_node` is more readable, I wonder if `cat_node` is more precise here, otherwise we rely on `list_node` not being `prim::FusedConcat` if `cat_node` is `aten::cat` (which certainly should be true, but isn't locally implied by the code) because that case isn't intended to be captured.

ngimel reviewed (commented):
Looks good to me, now that `where` is fused, would it also make sense to add its gradient to autodiff?

t-vi commented on torch/csrc/jit/fuser/codegen.cpp:
```diff
@@ -101,13 +149,13 @@ static std::string encodeRHS(const Node* n) {
     {aten::tanh, "tanhf(${0})"},
     {aten::sqrt, "sqrtf(${0})"},
-    {aten::rsqrt, "rsqrtf(${0})"},
+    {aten::rsqrt, "1.f/sqrtf(${0})"},
```
Ah no. I stumbled over `reciprocalf` and `rsqrtf` at the same time (on CPU) and erroneously extrapolated the *`reciprocalf` isn't a thing* assessment to `rsqrf`.
So we'd either define rsqrtf (along the lines of the Wikipedia article or just `1/sqrtf`?) on CPU manually? (Seems prettier than backend-specific code.)

t-vi committed:
Improvements based on review feedback

- Eliminate unneeded code,
- reinstate using rsqrtf in fuser, add rsqrtf to CPU template,
- add autodiff for max/min(Tensor a, Tensor b), where

Thank you, @ngimel for the comments!

t-vi committed:
Merge branch 'master' into fusion

t-vi committed:
Adapt to change in PropagateInputShapes signature

t-vi commented:
Thank you, @ngimel for the great comments! I added `where` and also `max(Tensor, Tensor)` and `min` to autodiff.

soumith reviewed (approved):
stamping to ship, reviewed by Natalia

apaszke commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -681,7 +685,7 @@ class ShapePropagator {
         [this](Node* node) -> type_vec_t {
           if (auto maybe_tensor_types = gatherTensorTypes<TensorType>(node)) {
-            return {broadcast(*maybe_tensor_types)};
+            return {broadcast(*maybe_tensor_types, (node->kind() == aten::where ? 1 : 0))};
```
Noooo, please move `aten::where` to another block instead.

apaszke commented on torch/csrc/jit/fuser/codegen.cpp:
```diff
@@ -77,10 +81,54 @@ static const char* scalarTypeName(const at::ScalarType type) {
+    auto const tt = t->cast<TensorType>();
+    return calcScalarTypeName(tt->scalarType());
+  }
+  // TODO: we might also error out if we think the above should cover everything
```
I think we really should assert that we can shape prop through fused kernels. They're very simple, and with the current code I can't really prove to myself that it's correct (mostly because of this `auto`).

apaszke reviewed (approved):
LGTM, but let's error out if we cannot figure out the explicit scalar type of an intermediate variable

t-vi committed:
Improvements from Adam's review comments

- Split where in shape analysis to avoid fudging with a separate
  test.
- Error out when the fuser codegen cannot determine the scalar type
  for an expression.

t-vi commented:
The CI failures seems to be a timeout (on macos) and a time out in a dataloader setup (`AssertionError: False is not true : loader process setup timed out`, on windows). I don't see the immediate connection to the PR.
